### ISSUE TITLE
More semantic highlighting using Slublime Text scope naming

### ISSUE THREAD

**Issue by amurzeau** [clangd/clangd#352]
Hi,

Currently, semantic highlighting is reported using textmate name scopes as defined here:
https://github.com/llvm-mirror/clang-tools-extra/blob/master/clangd/SemanticHighlighting.cpp#L491

To have more details on the IDE side, it would be intersting to have more scopes, like:
-  differentiate declaration from definition (functions and types)
- virtual functions
- goto labels
- maybe doxygen comments and tags (thought maybe it should belong to a doxygen language server instead)
- preprocessor defines
- type keywords which are reported as a type but IDE might want to display them as keywords instead

**Comment by sam-mccall** [comment]
We should add more details and token types.

Things that are easily lexically derived (such as preprocessor defines and doxygen tags) don't need to be in the language server I think. At least, we haven't tried so far to make the coloring exhaustive, because we assume it's always a layer over syntax highlighting. Compared to syntax highlighting, LSP based syntax highlighting is asynchronous and imposes extra costs like serialization.

Since you mention textmate scopes: LSP 3.16 has "official" support for semantic highlighting (which doesn't use textmate scopes) and clangd now supports this: https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clangd/SemanticHighlighting.h#L16. Future enhancements highlighting (in clangd or client integrations) should probably be based on the new model, we'll remove the older one eventually.

**Comment by amurzeau** [comment]
> 
> 
> We should add more details and token types.
> 
> Things that are easily lexically derived (such as preprocessor defines and doxygen tags) don't need to be in the language server I think. At least, we haven't tried so far to make the coloring exhaustive, because we assume it's always a layer over syntax highlighting. Compared to syntax highlighting, LSP based syntax highlighting is asynchronous and imposes extra costs like serialization.

I agree that some stuff might be better as a syntax highlighting on the client side while semantic only should be reported by the language server.
So things like keywords can be left to the client, but not much to me, as it would require every clients to implement non negligible parsers specific to the language
What does notepad++ and kate for syntax highlighting is fine for the client to me. Fast  and easy enough so the user doesn't have black and white text while the language server is working.

Anyway the current state with clangd seems ok to me about that.
Actually, preprocessor defines are handled by clangd, I should not have put it into my list in description.

> 
> Since you mention textmate scopes: LSP 3.16 has "official" support for semantic highlighting (which doesn't use textmate scopes) and clangd now supports this: https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clangd/SemanticHighlighting.h#L16. Future enhancements highlighting (in clangd or client integrations) should probably be based on the new model, we'll remove the older one eventually.

I did not know that and I agree this new model is what to target now, it seems better at handling modifiers like static or virtual for C++. This sounds more promising for better language semantic information to rely on :)

I see that token modifiers are not implemented, but I'm eager to see that :)

Thanks for your reply !

**Comment by amurzeau** [comment]
Also, an interesting information is a modifier to indicate if a parameter is passed by non const reference, which would allow the user to see where a variable can be modified.
For example, QtCreator:
![image](https://user-images.githubusercontent.com/5435069/80320177-28e4de80-8815-11ea-8140-93c0fa13f485.png)

and this web source code browser:
https://code.woboq.org/qt5/qtbase/src/widgets/widgets/qwidgettextcontrol.cpp.html#3001
![image](https://user-images.githubusercontent.com/5435069/80320141-e91df700-8814-11ea-8ff6-94bff6bd5bb6.png)

It does show implicit conversions too, but I don't see a good way to pass this information via LSP.

**Comment by HighCommander4** [comment]
> Also, an interesting information is a modifier to indicate if a parameter is passed by non const reference, which would allow the user to see where a variable can be modified.

+1, Eclipse CDT has [had this for a while](https://wiki.eclipse.org/CDT/User/NewIn92#Syntax_coloring_for_variables_passed_by_non-const_reference).

**Comment by HighCommander4** [comment]
> Also, an interesting information is a modifier to indicate if a parameter is passed by non const reference, which would allow the user to see where a variable can be modified.

Note, this has been implemented in #839.
